src/core/producerconsumerqueue.h

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CORE_PRODUCERCONSUMERQUEUE_H
+#define BITCOIN_CORE_PRODUCERCONSUMERQUEUE_H
+
+#include <assert.h>
+#include <deque>
+#include <sync.h>
+#include <type_traits>
+
+/**
+ * The mode in which the queue operates
+ * Modes may be specified for both producers and consumers
+ */
+enum class WorkerMode {
+    BLOCKING,   //!< cv_wait until the action may proceed
+    NONBLOCKING //!< do not block, immediately return failure if the action is not possible
+};
+
+/**
+ * A FIFO thread safe producer consumer queue with two operations Push() and Pop()
+ * Producers Push() and Consumers Pop()
+ *
+ * @param T the type of the data contained
+ * @param m_producer_mode queue behavior when calling Push() on a full queue (block till space becomes available, or immediately fail)
+ * @param m_consumer_mode queue behavior when calling Pop() on an empty queue (block until there is data, or immediately fail)
+ *
+ * @see WorkerMode
+ */
+template <typename T, WorkerMode m_producer_mode = WorkerMode::BLOCKING, WorkerMode m_consumer_mode = WorkerMode::BLOCKING>
```

Comment:
In commit "Implement a thread-safe FIFO (producer/consumer style) queue" (d8106146e7511837decb4b45de35dec654454b69)

I don't think it's a good idea for blocking and nonblocking defaults to be attributes of the Queue data structure, instead of arguments to (or variants of) the push and pop methods. Advantages to dropping these template arguments:

1) Readability. It would be nice to be able to see if a push or pop call is blocking just by looking at the call, without have to check another part of the code to see how the queue data structure was initially declared.
2) Code size. Dropping these arguments would avoid compiler potentially having to instantiate many copies of this code for different combinations of template arguments.
3) Extensibility. There could be other useful blocking methods added in the future (like methods to wait for low/high water marks or for empty/full events) and it would either be verbose to have to add new classwide blocking/nonblocking defaults for new methods, or really confusing to have to somehow tie existing defaults to new methods.
4) Consistency. If you look at other C++ objects that support optional blocking like std::mutex or std::future, the blocking behaviour is determined only by the particular method call, not by template arguments from where the object was declared.

Reply:
I can see this either way - i wrote it with defaults essentially as constructor args because - at least for the use cases i can imagine you almost always want a default mode of operation for a given queue except for certain edge cases (shutdown is the most apparent one) - and i've seen data structures that handle this sort of initialization both ways (defaults on construction vs with every method call) but i'm also happy to change this up if the prevailing opinion is the other way.

Reply:
> Readability. It would be nice to be able to see if a push or pop call is blocking just by looking at the call

I think this comment is spot on.

>  If you look at other C++ objects that support optional blocking like std::mutex or std::future, the blocking behaviour is determined only by the particular method call

> What about having Push, Pop, TryPush, and TryPop methods and dropping the enum entirely?

Agree.